bindings/node/src/mongocrypt.cc

```diff
@@ -230,8 +238,129 @@ void MongoCrypt::logHandler(mongocrypt_log_level_t level,
+
+            v8::Local<v8::Value> argv[] = {keyBuffer, ivBuffer, inBuffer, outBuffer};
+            v8::Local<v8::Value> defaultValue = Nan::False();
+            v8::Local<v8::Value> result =
+                Nan::Call(*hook, Nan::GetCurrentContext()->Global(), 4, argv).FromMaybe(defaultValue);
+
+            if (!result->IsNumber()) {
+                // TODO: error checking, use status
+                return false;
+            }
+
+            *bytes_written = result->Uint32Value();
+            return true;
+        };
+
+    auto random =
+        [](void *ctx, mongocrypt_binary_t *out, uint32_t count, mongocrypt_status_t *status) -> bool{
+            Nan::HandleScope scope;
+            CryptoHooks* cryptoHooks = static_cast<CryptoHooks*>(ctx);
+            Nan::Callback* hook = cryptoHooks->randomHook.get();
+
+            v8::Local<v8::Object> outBuffer = BufferFromBinaryNoCopy(out);
+            v8::Local<v8::Value> argv[] = {outBuffer, Nan::New(count)};
+            Nan::Call(*hook, Nan::GetCurrentContext()->Global(), 2, argv);
+            return true;
+        };
+
+    auto hmac_sha_512 =
+        [](void *ctx, mongocrypt_binary_t *key, mongocrypt_binary_t *in, mongocrypt_binary_t *out, mongocrypt_status_t *status) -> bool {
+            Nan::HandleScope scope;
+            CryptoHooks* cryptoHooks = static_cast<CryptoHooks*>(ctx);
+            Nan::Callback* hook = cryptoHooks->hmacSha512Hook.get();
+
+            v8::Local<v8::Object> keyBuffer = BufferFromBinary(key);
+            v8::Local<v8::Object> inputBuffer = BufferFromBinary(in);
+            v8::Local<v8::Object> outputBuffer = BufferFromBinaryNoCopy(out);
+
+            // TODO: error checking, use status if passed back value is not undefined
+            v8::Local<v8::Value> argv[] = {keyBuffer, inputBuffer, outputBuffer};
+            Nan::Call(*hook, Nan::GetCurrentContext()->Global(), 3, argv);
+            return true;
+        };
+
+    auto hmac_sha_256 =
+        [](void *ctx, mongocrypt_binary_t *key, mongocrypt_binary_t *in, mongocrypt_binary_t *out, mongocrypt_status_t *status) -> bool {
+            Nan::HandleScope scope;
+            CryptoHooks* cryptoHooks = static_cast<CryptoHooks*>(ctx);
+            Nan::Callback* hook = cryptoHooks->hmacSha256Hook.get();
+
+            v8::Local<v8::Object> keyBuffer = BufferFromBinary(key);
+            v8::Local<v8::Object> inputBuffer = BufferFromBinary(in);
+            v8::Local<v8::Object> outputBuffer = BufferFromBinaryNoCopy(out);
+
+            // TODO: error checking, use status if passed back value is not undefined
+            v8::Local<v8::Value> argv[] = {keyBuffer, inputBuffer, outputBuffer};
+            Nan::Call(*hook, Nan::GetCurrentContext()->Global(), 3, argv);
+            return true;
+        };
+
+    auto sha_256 =
+        [](void *ctx, mongocrypt_binary_t *in, mongocrypt_binary_t *out, mongocrypt_status_t *status) -> bool {
```

Comment:
to understand the workflow of these: when `libmongocrypt` calls these functions, the first parameter is going to be the original `CryptoHooks` struct that we passed into `mongocrypt_setopt_crypto_hooks`, so we just have to cast it back out?

Reply:
yep! it's a common pattern with C api's that you allow the caller to pass through `void*` data, so that you can use your own types in the callback. Particularly in this case its needed because C++ doesn't allow lambdas to capture in these situations.